waspc/src/Wasp/Generator/Common.hs

```diff
@@ -11,15 +13,36 @@ import Text.Printf (printf)
 -- | Directory where the whole web app project (client, server, ...) is generated.
 data ProjectRootDir
 
+type SemanticVersion = (Int, Int, Int)
```

Comment:
Ha, you went all in for SemanticVersion :D!

As @shayneczyzewski said, I think it is ok to use `type` to document stuff additionally. The main question becomes, do you want to go step further and introduce proper data type? Sometimes answer will be no, for practical reasons, because you don't want to deal with all the conversions and similar.

Here however, it seems to me you already introduced decent amount of complexity, so maybe it is a good moment to introduce a `newtype` instead of just `type`.

But some other observations first: `compatibleWith :: String` feels very out of context, looking just at its signature. And you have a lot of these functions that are all about semver, but are in Wasp.Generator.Common. So for start, I would take them out into Wasp.Semver (create it).

Then, I would consider naming. So you have Version vs SemverString. Thinking about them, what is the difference? Version is really SemverVersion. So we have SemverVersion vs SemverString. What is the difference? Is it also version but as a string? Or is it smth else?
I would suggest different terminology: Version and VersionBounds. Version is `1.2.3`, VersionBounds is `>= 1.2.3` or `== 1.2.3` or `^1.2.3`. I have to admit I am mostly getting the idea for this from cabal(PVP) dependency terminology, so maybe it is worth exploring more what is the terminology that npm uses.

You could go pretty fancy with types here: introduce `data Version`, then maybe `data VersionBounds`, and maybe even a `data VersionBoundOperator`, maybe even `data VersionBound`. Or maybe just part of this.

I think this will also become clearer once you extract it into a separate file and play with it there a bit.

Reply:
> Here however, it seems to me you already introduced decent amount of complexity, so maybe it is a good moment to introduce a newtype instead of just type.
>
> But some other observations first: compatibleWith :: String feels very out of context, looking just at its signature. And you have a lot of these functions that are all about semver, but are in Wasp.Generator.Common. So for start, I would take them out into Wasp.Semver (create it).
> [...]
> I think this will also become clearer once you extract it into a separate file and play with it there a bit.

Sure, I'll clean it up a bit.

> Then, I would consider naming. So you have Version vs SemverString. Thinking about them, what is the difference? Version is really SemverVersion. So we have SemverVersion vs SemverString. What is the difference? Is it also version but as a string? Or is it smth else?

The idea was this:
- A `Version` is a version identifier consisting of a major, minor and patch value
- A `SemverString` is a string that specifies the compatibility. It can just be the version or have some additional symbols.
- 
So yes, `VersionBounds` or `VersionSpecifier` definitely make more sense, I'll change it.

Reply:
Maybe good naming is Version, VersionBounds, and BoundsSpecifier.
Version is a.b.c, VersionBounds is an expressions consisting of Versions and BoundSpecifiers, i.e. >= 1.2 && < 2.3, while VersionSpecifier is >= or <. But it can also be 1.* right? Ah upf ok in that case it is not so simple. Maybe best to just have Version vs VersionSpecifier / Bounds and that is it, not complicate it too much. Just thinking out loud, I am sure you will figure out something reasonable, probably best not to overengineer.